support containerLogMax[Size|Files] kubelet configurations (#6702)
gardener/gardener: docs/usage/logging.md

* One Loki Statefulset in the `garden` namespace which contains logs for the seed cluster and one per shoot namespace which contains logs for shoot's controlplane.
* One Grafana Deployment in `garden` namespace and two Deployments per shoot namespace (one exposed to the end users and one for the operators). Grafana is the UI component used in the logging stack.

### Container Logs rotation and retention

> Container [Log rotation](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) in kubernetes describes a subtile but important implementation detail depending on the type of used high-level container runtime. When high level container runtime is used such as `docker shim`, kubernetes components do not provide any rotation or retention implementations leaving those aspects to the downstream components. When `containerd` a cri-based container runtime is used then the Kubelet provides the necessary implementation with two configuration options: `ContainerLogMaxSize` for rotation and `ContainerLogMaxFiles` for retention.

Reviewer: ```suggestion

Container [log rotation](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) in Kubernetes describes a subtile but important implementation detail depending on the type of the used high-level container runtime. When a high-level container runtime is used (such as `docker shim`) then Kubernetes components do not provide any rotation or retention implementations, hence leaving those aspects to the downstream components. When `containerd` (a CRI-based container runtime) is used then the `kubelet` provides the necessary implementation with two configuration options:
- `ContainerLogMaxSize` for rotation
- `ContainerLogMaxFiles` for retention.
```

#### Docker container runtime

> In this case the log rotation and retention is implemented by a logrotate service provisioned by gardener which rotates logs once 100M size is reached. Logs are compressed on daily basis and retained for a maximum period of 14 days.

Reviewer: ```suggestion

In this case, the log rotation and retention is implemented by a `logrotate` service provisioned by Gardener which rotates logs once `100M` size is reached. Logs are compressed on daily basis and retained for a maximum period of `14d`.
```

#### ContainerD runtime

> In this case, the Shoot may supply `ContainerLogMaxSize` and `ContainerLogMaxFiles`. Both are optional and if nothing is specified then the

Reviewer: ```suggestion

In this case, the `Shoot` specification may contain `ContainerLogMaxSize` and `ContainerLogMaxFiles`. Both are optional and if nothing is specified then the
```

> kubelet rotate on the same size `100M` as in the docker container runtime.

Reviewer: ```suggestion
kubelet rotates on the same size `100M` as in the docker container runtime.

```

```yaml

> kubernetes:

Reviewer: ```suggestion
spec:
  kubernetes:
```

    kubelet:
      # accepted values are of resource.Quantity
      containerLogMaxSize: 150Mi
      containerLogMaxFiles: 10
```

> Those values of the `ContainerLogMaxSize` and `ContainerLogMaxFiles` properties needs to be considered with care since container log files claims disk space from the host. On the oposite side ,log rotations on sizes too small may result in frequent rotations which can be missed by other components (log shippers) observing these rotations.

Reviewer: ```suggestion
Those values of the `ContainerLogMaxSize` and `ContainerLogMaxFiles` properties need to be considered with care since container log files claim disk space from the host. On the opposite side, log rotations on sizes too small may result in frequent rotations which can be missed by other components (log shippers) observing these rotations.
```

> In the majority of the cases the defaults shall do just. Custom configuration might be of use under rare conditions.

Reviewer: ```suggestion
In the majority of the cases, the defaults shall do just. Custom configuration might be of use under rare conditions.
```

* One Loki Statefulset in the `garden` namespace which contains logs for the seed cluster and one per shoot namespace which contains logs for shoot's controlplane.
* One Grafana Deployment in `garden` namespace and two Deployments per shoot namespace (one exposed to the end users and one for the operators). Grafana is the UI component used in the logging stack.

> ### Container Logs rotation and retention

Reviewer 2: I am not sure that this is the right place for the newly added docs.
What is the target audience of this doc? I think it is rather a developer/operator but not a Shoot owner/end user (the guide explains internals of loki, fluent-bit what configurations they use and this is obviously makes it not applicable for Shoot owners). WDYT think about moving this doc to a new guide (`docs/usage/container-log-rotation`) with a clear target audience - Shoot owner/end user?

Author: No it is not. This the correct place.

> Container [log rotation](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) in Kubernetes describes a subtile but important implementation detail depending on the type of the used high-level container runtime. When a high-level container runtime is used (such as `docker shim`) then Kubernetes components do not provide any rotation or retention implementations, hence leaving those aspects to the downstream components. When `containerd` (a CRI-based container runtime) is used then the `kubelet` provides the necessary implementation with two configuration options:

Reviewer 2: s/docker shim/dockershim/

> Container [log rotation](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) in Kubernetes describes a subtile but important implementation detail depending on the type of the used high-level container runtime. When a high-level container runtime is used (such as `docker shim`) then Kubernetes components do not provide any rotation or retention implementations, hence leaving those aspects to the downstream components. When `containerd` (a CRI-based container runtime) is used then the `kubelet` provides the necessary implementation with two configuration options:

Reviewer 2: It is rather about whether the container runtime is CRI compliant or not (and not "high-level" or not).

```suggestion
Container [log rotation](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) in Kubernetes describes a subtile but important implementation detail depending on the type of the used high-level container runtime. When the used container runtime is not CRI compliant (such as `dockershim`) then the `kubelet` does not provide any rotation or retention implementations, hence leaving those aspects to the downstream components. When the used container runtime is CRI compliant (such as `containerd`) then the `kubelet` provides the necessary implementation with two configuration options:
```

- `ContainerLogMaxSize` for rotation
- `ContainerLogMaxFiles` for retention.

#### Docker container runtime

> In this case, the log rotation and retention is implemented by a `logrotate` service provisioned by Gardener which rotates logs once `100M` size is reached. Logs are compressed on daily basis and retained for a maximum period of `14d`.

Reviewer 2: > Logs are compressed on daily basis and retained for a maximum period of `14d`.

This is also true with `containerd`, right? Because written like this, my understanding as a reader is that this daily rotation by `logrotate` happens only for `docker`. Can you rework the section to make it clear that the above statement is also true when `containerd` is used?

Author: it is clear enough

#### ContainerD runtime

> In this case, the `Shoot` specification may contain `ContainerLogMaxSize` and `ContainerLogMaxFiles`. Both are optional and if nothing is specified then the

Reviewer 2: ```suggestion
In this case, it is possible to configure the `containerLogMaxSize` and `containerLogMaxFiles` fields in the Shoot specification. Both fields are optional and if nothing is specified then the `kubelet` rotates on the same size `100M` as in the `docker` container runtime.
```

kubelet rotates on the same size `100M` as in the docker container runtime.

```yaml
spec:
  kubernetes:
    kubelet:
      # accepted values are of resource.Quantity
      containerLogMaxSize: 150Mi
      containerLogMaxFiles: 10
```

> Those values of the `ContainerLogMaxSize` and `ContainerLogMaxFiles` properties need to be considered with care since container log files claim disk space from the host. On the opposite side, log rotations on sizes too small may result in frequent rotations which can be missed by other components (log shippers) observing these rotations.

Reviewer 2: ```suggestion
The values of the `containerLogMaxSize` and `containerLogMaxFiles` fields need to be considered with care since container log files claim disk space from the host. On the opposite side, log rotations on too small sizes may result in frequent rotations which can be missed by other components (log shippers) observing these rotations.
```
